File: Engine/Android/core/src/test/kotlin/despia/engine/FormsConformanceTest.kt
package despia.engine

import java.io.File
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * The U08 forms conformance runner - executes
 * OpenSource/Conformance/forms/{mask,countries,phone,daterange,validation,composites}.json
 * through THIS runtime's Forms core (parity/U08-forms.md). The TS reference
 * (packages/kernel/test/forms-conformance.test.ts) and the Swift twin (record lane,
 * FormsConformance.swift) run the SAME files, so "where is the caret after an edit in the
 * middle of a masked value" cannot have three answers, a phone number cannot be E.164 on one
 * renderer and national on the next, and a double-tapped submit cannot be blocked on two
 * platforms and armed on the third.
 *
 * Missing corpus = loud failure: a silently-skipped conformance suite is how drift starts.
 */
class FormsConformanceTest {

    private fun corpusDir(): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/forms")
            if (candidate.isDirectory) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/forms not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun doc(name: String): Map<String, Any?> {
        val file = File(corpusDir(), "$name.json")
        assertTrue(file.isFile, "$name.json missing")
        return json(file.readText()).foundationValue as? Map<String, Any?>
            ?: error("$name.json: not a JSON object")
    }

    @Suppress("UNCHECKED_CAST")
    private fun rows(name: String, section: String): List<Map<String, Any?>> {
        val list = doc(name)[section] as? List<Map<String, Any?>> ?: error("$name.json: no $section[]")
        assertTrue(list.isNotEmpty(), "$name.$section must not be empty")
        return list
    }

    private fun int(value: Any?): Int = (value as Number).toInt()

    private fun str(value: Any?): String = value as? String ?: ""

    @Suppress("UNCHECKED_CAST")
    private fun strings(value: Any?): List<String> =
        (value as? List<Any?> ?: emptyList()).map { it as String }

    // ── countries.json - ONE table, asserted row for row against the shipped literal ──────

    @Test
    @Suppress("UNCHECKED_CAST")
    fun theKernelTableIsTheCorpusTable() {
        val root = doc("countries")
        assertEquals(root["default"], Forms.DEFAULT_COUNTRY, "default country")
        val wanted = root["countries"] as List<Map<String, Any?>>
        assertEquals(wanted.size, Forms.COUNTRIES.size, "country count")
        assertTrue(Forms.COUNTRIES.size >= 60, "the country table is suspiciously small")
        wanted.forEachIndexed { index, want ->
            val got = Forms.COUNTRIES[index]
            val label = "country row $index (${want["iso"]})"
            assertEquals(want["iso"], got.iso, "$label: iso")
            assertEquals(want["name"], got.name, "$label: name")
            assertEquals(want["dial"], got.dial, "$label: dial")
            assertEquals(want["trunk"], got.trunk, "$label: trunk")
            assertEquals(int(want["nsnMin"]), got.nsnMin, "$label: nsnMin")
            assertEquals(int(want["nsnMax"]), got.nsnMax, "$label: nsnMax")
            assertEquals(want["format"], got.format, "$label: format")
            assertEquals(want["primary"], got.primary, "$label: primary")
        }
    }

    @Test
    fun aNationalFormatNeverTruncatesALegalNumber() {
        for (row in Forms.COUNTRIES) {
            val format = row.format ?: continue
            assertEquals(row.nsnMin, row.nsnMax, "${row.iso}: a format needs a fixed NSN length")
            assertEquals(row.nsnMin, Forms.maskCapacity(format), "${row.iso}: format capacity != NSN length")
        }
    }

    @Test
    fun everyIsoIsUniqueAndResolvable() {
        val seen = HashSet<String>()
        for (row in Forms.COUNTRIES) {
            assertTrue(seen.add(row.iso), "duplicate ISO ${row.iso}")
            assertEquals(row.dial, Forms.country(row.iso)?.dial, "${row.iso}: uppercase lookup")
            assertEquals(row.dial, Forms.country(row.iso.lowercase())?.dial, "${row.iso}: lowercase lookup")
        }
    }

    // ── mask.json ─────────────────────────────────────────────────────────────────────────

    @Test
    fun maskCapacityAndDescriptionAgreeWithCorpus() {
        for (case in rows("mask", "capacity")) {
            val mask = str(case["mask"])
            assertEquals(int(case["capacity"]), Forms.maskCapacity(mask), "capacity of \"$mask\"")
            assertEquals(case["description"], Forms.maskDescription(mask), "description of \"$mask\"")
        }
    }

    @Test
    fun maskFormatAgreesWithCorpus() {
        for (case in rows("mask", "format")) {
            val mask = str(case["mask"])
            val raw = str(case["raw"])
            val display = Forms.maskFormat(mask, raw)
            assertEquals(case["display"], display, "format \"$mask\" <- \"$raw\"")
            assertEquals(case["extract"], Forms.maskExtract(mask, display), "extract of the display")
        }
    }

    @Test
    fun maskExtractAgreesWithCorpus() {
        for (case in rows("mask", "extract")) {
            val mask = str(case["mask"])
            val text = str(case["text"])
            assertEquals(case["raw"], Forms.maskExtract(mask, text), "extract \"$mask\" <- \"$text\"")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun theCaretLawAgreesWithCorpus() {
        val cases = rows("mask", "edits")
        assertTrue(cases.size >= 20, "forms/mask edits corpus is suspiciously small")
        for (case in cases) {
            val name = str(case["name"])
            val mask = str(case["mask"])
            val got = Forms.maskEdit(
                mask, str(case["prev"]), int(case["selStart"]), int(case["selEnd"]), str(case["insert"]),
            )
            val expect = case["expect"] as Map<String, Any?>
            assertEquals(expect["display"], got.display, "$name: display")
            assertEquals(int(expect["caret"]), got.caret, "$name: caret")
            assertEquals(expect["raw"], got.raw, "$name: raw")
            assertEquals(expect["complete"], got.complete, "$name: complete")
            // THE PLACEMENT LAW: bind can never hold a character the field does not show.
            assertEquals(got.raw, Forms.maskExtract(mask, got.display), "$name: display and raw disagree")
        }
    }

    // ── phone.json ────────────────────────────────────────────────────────────────────────

    @Test
    fun theFlagIsAPureFunctionOfTheIsoCode() {
        for (case in rows("phone", "flags")) {
            val iso = str(case["iso"])
            assertEquals(case["flag"], Forms.flag(iso), "flag $iso")
        }
        assertEquals("", Forms.flag("USA"), "a three-letter code is not an ISO alpha-2")
        assertEquals("", Forms.flag(""), "the empty code has no flag")
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun e164ParsingAgreesWithCorpus() {
        val cases = rows("phone", "parse")
        assertTrue(cases.size >= 30, "forms/phone parse corpus is suspiciously small")
        for (case in cases) {
            val input = str(case["input"])
            val label = "\"$input\" @ ${case["defaultCountry"] ?: "-"}"
            val got = Forms.phoneParse(input, case["defaultCountry"] as? String)
            val expect = case["expect"] as Map<String, Any?>
            assertEquals(expect["e164"], got.e164, "$label: e164")
            assertEquals(expect["national"], got.national, "$label: national")
            assertEquals(expect["country"], got.country, "$label: country")
            assertEquals(expect["dialCode"], got.dialCode, "$label: dialCode")
            assertEquals(expect["nsn"], got.nsn, "$label: nsn")
            assertEquals(expect["valid"], got.valid, "$label: valid")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun aValidParseRoundTripsThroughItsOwnE164() {
        for (case in rows("phone", "parse")) {
            val expect = case["expect"] as Map<String, Any?>
            if (expect["valid"] != true) continue
            val e164 = str(expect["e164"])
            val again = Forms.phoneParse(e164, null)
            assertEquals(e164, again.e164, "${case["input"]} did not round-trip")
            assertTrue(again.valid, "${case["input"]} lost validity on round-trip")
        }
    }

    // ── daterange.json ────────────────────────────────────────────────────────────────────

    @Test
    @Suppress("UNCHECKED_CAST")
    fun theMonthGridAgreesWithCorpus() {
        for (case in rows("daterange", "grid")) {
            val year = int(case["year"])
            val month = int(case["month"])
            val got = Forms.monthGrid(year, month, int(case["firstWeekday"]))
            val expect = case["expect"] as Map<String, Any?>
            val label = "$year-$month fw${case["firstWeekday"]}"
            assertEquals(int(expect["days"]), got.days, "$label: days")
            assertEquals(int(expect["leading"]), got.leading, "$label: leading")
            assertEquals(int(expect["weeks"]), got.weeks, "$label: weeks")
        }
    }

    @Test
    fun civilDayNumbersAgreeWithCorpusAndRoundTrip() {
        for (case in rows("daterange", "dayNumbers")) {
            val date = str(case["date"])
            val day = case["day"]
            if (day == null) {
                assertNull(Forms.dayNumber(date), "$date must not resolve")
                continue
            }
            assertEquals(int(day), Forms.dayNumber(date), "dayNumber $date")
            assertEquals(date, Forms.dateFromDay(int(day)), "dateFromDay ${int(day)}")
        }
    }

    @Test
    fun aDstTransitionDayIsExactlyOneDayLong() {
        for (case in rows("daterange", "dstAdjacency")) {
            val from = Forms.dayNumber(str(case["from"])) ?: error("bad from")
            val to = Forms.dayNumber(str(case["to"])) ?: error("bad to")
            assertEquals(int(case["delta"]), to - from, "${case["from"]} -> ${case["to"]}")
            assertEquals(case["next"], Forms.dateFromDay(from + 1), "the day after ${case["from"]}")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun selectabilityAgreesWithCorpus() {
        for (case in rows("daterange", "selectable")) {
            val config = foldConfig(case["config"] as? Map<String, Any?> ?: emptyMap())
            val date = str(case["date"])
            val got = Forms.selectable(config, date)
            val expect = case["expect"] as Map<String, Any?>
            assertEquals(expect["ok"], got.ok, "selectable $date: ok")
            assertEquals(expect["reason"], got.reason, "selectable $date: reason")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun foldConfig(raw: Map<String, Any?>): Forms.DateFoldConfig = Forms.DateFoldConfig(
        range = raw["range"] == true,
        min = raw["min"] as? String,
        max = raw["max"] as? String,
        disabledDates = strings(raw["disabledDates"]),
        start = raw["start"] as? String,
        end = raw["end"] as? String,
        month = raw["month"] as? String ?: "",
    )

    @Test
    @Suppress("UNCHECKED_CAST")
    fun theRangeFoldAgreesWithCorpus() {
        val folds = rows("daterange", "folds")
        assertTrue(folds.size >= 10, "forms/daterange fold corpus is suspiciously small")
        for (case in folds) {
            val name = str(case["name"])
            val config = foldConfig(case["config"] as? Map<String, Any?> ?: emptyMap())
            val events = (case["events"] as List<Map<String, Any?>>).map {
                Forms.DateFoldEvent(str(it["kind"]), str(it["date"]), str(it["month"]))
            }
            val steps = Forms.dateFold(config, events)
            val expect = case["expect"] as List<Map<String, Any?>>
            assertEquals(expect.size, steps.size, "$name: step count")
            expect.forEachIndexed { index, want ->
                val got = steps[index]
                assertEquals(want["start"], got.start, "$name step $index: start")
                assertEquals(want["end"], got.end, "$name step $index: end")
                assertEquals(want["month"], got.month, "$name step $index: month")
                assertEquals(want["complete"], got.complete, "$name step $index: complete")
                assertEquals(want["reason"], got.reason, "$name step $index: reason")
                assertEquals(strings(want["fired"]), got.fired, "$name step $index: fired")
            }
        }
    }

    // ── validation.json ───────────────────────────────────────────────────────────────────

    @Test
    fun theRuleVocabularyAgreesWithCorpus() {
        val cases = rows("validation", "rules")
        assertTrue(cases.size >= 15, "forms/validation rules corpus is suspiciously small")
        for (case in cases) {
            val rule = str(case["rule"])
            val got = Forms.rule(rule, str(case["arg"]), str(case["value"]), str(case["pattern"]))
            assertEquals(case["expect"], got, "$rule(${case["arg"]}) <- \"${case["value"]}\"")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun field(raw: Map<String, Any?>): Forms.FieldState = Forms.FieldState(
        name = str(raw["name"]),
        value = str(raw["value"]),
        initial = str(raw["initial"]),
        validate = str(raw["validate"]),
        pattern = str(raw["pattern"]),
        message = str(raw["message"]),
    )

    @Test
    @Suppress("UNCHECKED_CAST")
    fun theFirstFailingRuleOwnsTheMessage() {
        rows("validation", "fieldErrors").forEachIndexed { index, case ->
            val got = Forms.fieldError(field(case["field"] as Map<String, Any?>))
            assertEquals(case["expect"], got, "fieldError $index")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun formAggregationAgreesWithCorpus() {
        for (case in rows("validation", "aggregate")) {
            val name = str(case["name"])
            val fields = (case["fields"] as List<Map<String, Any?>>).map { field(it) }
            val got = Forms.aggregate(fields)
            val expect = case["expect"] as Map<String, Any?>
            assertEquals(expect["valid"], got.valid, "$name: valid")
            assertEquals(expect["dirty"], got.dirty, "$name: dirty")
            assertEquals(expect["errors"], got.errors, "$name: errors")
            assertEquals(strings(expect["invalid"]), got.invalid, "$name: invalid")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun theDoubleSubmitLawAgreesWithCorpus() {
        for (case in rows("validation", "submit")) {
            val name = str(case["name"])
            val state = case["state"] as Map<String, Any?>
            val fields = (case["fields"] as List<Map<String, Any?>>).map { field(it) }
            val got = Forms.submit(state["submitting"] == true, state["disabled"] == true, fields)
            val expect = case["expect"] as Map<String, Any?>
            assertEquals(expect["action"], got.action, "$name: action")
            assertEquals(expect["reason"], got.reason, "$name: reason")
            assertEquals(expect["submitting"], got.submitting, "$name: submitting")
            assertEquals(expect["disabled"], got.disabled, "$name: disabled")
            if (expect.containsKey("fields")) {
                assertEquals(strings(expect["fields"]), got.fields, "$name: fields")
            }
            if (expect.containsKey("submitted")) {
                assertEquals(expect["submitted"], got.submitted, "$name: submitted")
            }
            if (expect.containsKey("touchedAll")) {
                assertEquals(expect["touchedAll"], got.touchedAll, "$name: touchedAll")
            }
        }
    }

    // ── composites.json ───────────────────────────────────────────────────────────────────

    @Test
    @Suppress("UNCHECKED_CAST")
    fun multiSelectTogglingAgreesWithCorpus() {
        rows("composites", "multiSelect").forEachIndexed { index, case ->
            val got = Forms.multiSelectToggle(strings(case["selected"]), str(case["value"]), int(case["max"]))
            val expect = case["expect"] as Map<String, Any?>
            assertEquals(strings(expect["selected"]), got.selected, "multiSelect $index: selected")
            assertEquals(expect["changed"], got.changed, "multiSelect $index: changed")
            assertEquals(expect["reason"], got.reason, "multiSelect $index: reason")
        }
    }

    @Test
    fun theSelectionAnnouncementAgreesWithCorpus() {
        for (case in rows("composites", "multiSelectAnnouncement")) {
            val got = Forms.multiSelectAnnouncement(int(case["count"]), int(case["max"]))
            assertEquals(case["expect"], got, "announcement ${case["count"]}/${case["max"]}")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun tagAdditionAgreesWithCorpus() {
        rows("composites", "tagsAdd").forEachIndexed { index, case ->
            val raw = case["config"] as? Map<String, Any?> ?: emptyMap()
            val config = Forms.TagsConfig(
                separator = raw["separator"] as? String ?: "",
                max = (raw["max"] as? Number)?.toInt() ?: 0,
                validate = raw["validate"] as? String ?: "",
            )
            val got = Forms.tagsAdd(strings(case["tags"]), str(case["text"]), config)
            val expect = case["expect"] as Map<String, Any?>
            assertEquals(strings(expect["tags"]), got.tags, "tagsAdd $index: tags")
            assertEquals(strings(expect["added"]), got.added, "tagsAdd $index: added")
            val rejected = expect["rejected"] as List<Map<String, Any?>>
            assertEquals(rejected.size, got.rejected.size, "tagsAdd $index: rejection count")
            rejected.forEachIndexed { at, want ->
                assertEquals(want["tag"], got.rejected[at].tag, "tagsAdd $index: rejection $at tag")
                assertEquals(want["reason"], got.rejected[at].reason, "tagsAdd $index: rejection $at reason")
            }
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun backspaceOnAnEmptyQueryRemovesTheLastChip() {
        rows("composites", "tagsBackspace").forEachIndexed { index, case ->
            val got = Forms.tagsBackspace(strings(case["tags"]), str(case["query"]))
            val expect = case["expect"] as Map<String, Any?>
            assertEquals(strings(expect["tags"]), got.tags, "tagsBackspace $index: tags")
            assertEquals(expect["removed"], got.removed, "tagsBackspace $index: removed")
        }
    }

    @Test
    fun theTagAnnouncementAgreesWithCorpus() {
        for (case in rows("composites", "tagsAnnouncement")) {
            val got = Forms.tagsAnnouncement(str(case["kind"]), str(case["tag"]), int(case["count"]))
            assertEquals(case["expect"], got, "tag announcement ${case["kind"]} ${case["tag"]}")
        }
    }

    // ── the laws that are not a corpus row but must never regress ─────────────────────────

    @Test
    fun aMaskedFieldNeverBindsWhatItDoesNotShow() {
        val got = Forms.maskEdit("AA-####", "", 0, 0, "12ab34")
        assertEquals("ab-34", got.display)
        assertEquals("ab34", got.raw)
        assertTrue(!got.complete, "four of six slots are filled")
        // and the field is still typable: the rejected digits did not eat its capacity
        val next = Forms.maskEdit("AA-####", got.display, got.caret, got.caret, "5")
        assertEquals("ab-345", next.display)
    }

    @Test
    fun anUnknownDialCodeIsATypedFailureNeverAGuess() {
        val got = Forms.phoneParse("+9991234567", null)
        assertNull(got.country, "no country")
        assertTrue(!got.valid, "not valid")
        assertEquals("+9991234567", got.e164, "the digits survive for the caller to show")
    }

    @Test
    fun requiredIsTheOnlyRuleAnEmptyValueCanFail() {
        assertTrue(!Forms.rule("required", "", "", ""))
        for (name in listOf("email", "url", "phone", "minLength", "maxLength", "pattern")) {
            assertTrue(Forms.rule(name, "8", "", "^x$"), "$name must pass on empty")
        }
    }

    @Test
    fun aRangeContainingADisabledDateIsRefusedWhole() {
        val config = Forms.DateFoldConfig(
            range = true, disabledDates = listOf("2026-03-11"), month = "2026-03",
        )
        val steps = Forms.dateFold(
            config,
            listOf(
                Forms.DateFoldEvent("select", "2026-03-09"),
                Forms.DateFoldEvent("select", "2026-03-13"),
            ),
        )
        assertEquals("range-contains-disabled", steps[1].reason)
        assertNull(steps[1].end, "the close was refused, not clipped")
    }

    @Test
    fun theSubmitGateBlocksTheSecondTapOfADoubleTap() {
        val fields = listOf(Forms.FieldState(name = "email", value = "ada@example.com", validate = "required,email"))
        val first = Forms.submit(false, false, fields)
        assertEquals("submit", first.action)
        assertTrue(first.submitting, "the first submit arms the flag")
        val second = Forms.submit(first.submitting, false, fields)
        assertEquals("blocked", second.action)
        assertEquals("in-flight", second.reason)
    }
}

File: Web/packages/kernel/test/forms-conformance.test.ts
//
//  forms-conformance.test.ts — the SHARED FORMS corpus
//  (OpenSource/Conformance/forms/{mask,countries,phone,daterange,validation,composites}
//  .json) through the TS forms kernel — the REFERENCE leg of U08. The Kotlin twin
//  (:core FormsConformanceTest) and the Swift twin (record lane, FormsConformance) run
//  the SAME files, so "where is the caret after an edit in the middle of a masked value"
//  cannot have three answers. Expected values were computed by an independent scratch
//  implementation, never by this kernel.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  maskCapacity, maskDescription, maskEdit, maskExtract, maskFormat,
  FORMS_COUNTRIES, FORMS_DEFAULT_COUNTRY, formsCountry, formsFlag, phoneParse,
  formsDayNumber, formsDateFromDay, formsMonthGrid, formsSelectable, formsDateFold,
  formsRule, formsFieldError, formsAggregate, formsSubmit,
  multiSelectToggle, multiSelectAnnouncement, tagsAdd, tagsBackspace, tagsAnnouncement,
  type DateFoldEvent, type FormFieldState,
} from "../src/forms.ts";

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/forms");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("forms corpus not found");
    dir = parent;
  }
}

function load<T>(file: string): T {
  return JSON.parse(readFileSync(join(corpusDir(), file), "utf-8")) as T;
}

// ── countries.json — ONE table, asserted byte-for-byte against the shipped literal ────

type CountryRow = {
  iso: string; name: string; dial: string; trunk: string;
  nsnMin: number; nsnMax: number; format: string | null; primary: boolean;
};
const countriesDoc = load<{ default: string; countries: CountryRow[] }>("countries.json");

test("forms/countries — the kernel table IS the corpus table", () => {
  assert.equal(FORMS_DEFAULT_COUNTRY, countriesDoc.default);
  assert.equal(FORMS_COUNTRIES.length, countriesDoc.countries.length);
  assert.ok(FORMS_COUNTRIES.length >= 60, "the country table is suspiciously small");
  for (let i = 0; i < countriesDoc.countries.length; i += 1) {
    const want = countriesDoc.countries[i]!;
    const got = FORMS_COUNTRIES[i]!;
    assert.deepEqual(
      { iso: got.iso, name: got.name, dial: got.dial, trunk: got.trunk,
        nsnMin: got.nsnMin, nsnMax: got.nsnMax, format: got.format, primary: got.primary },
      want, `country row ${i} (${want.iso})`,
    );
  }
});

test("forms/countries — a national format never truncates a legal number", () => {
  for (const row of countriesDoc.countries) {
    if (row.format === null) continue;
    assert.equal(row.nsnMin, row.nsnMax, `${row.iso}: a format needs a fixed NSN length`);
    assert.equal(maskCapacity(row.format), row.nsnMin, `${row.iso}: format capacity != NSN length`);
  }
});

test("forms/countries — every ISO is unique and resolvable", () => {
  const seen = new Set<string>();
  for (const row of countriesDoc.countries) {
    assert.ok(!seen.has(row.iso), `duplicate ISO ${row.iso}`);
    seen.add(row.iso);
    assert.equal(formsCountry(row.iso)?.dial, row.dial);
    assert.equal(formsCountry(row.iso.toLowerCase())?.dial, row.dial);
  }
});

// ── mask.json ─────────────────────────────────────────────────────────────────────────

type MaskDoc = {
  capacity: { mask: string; capacity: number; description: string }[];
  format: { mask: string; raw: string; display: string; extract: string }[];
  extract: { mask: string; text: string; raw: string }[];
  edits: {
    name: string; mask: string; prev: string; selStart: number; selEnd: number; insert: string;
    expect: { display: string; caret: number; raw: string; complete: boolean };
  }[];
};
const maskDoc = load<MaskDoc>("mask.json");
assert.ok(maskDoc.edits.length >= 20, "forms/mask edits corpus is suspiciously small");

for (const c of maskDoc.capacity) {
  test(`forms/mask capacity — ${JSON.stringify(c.mask)}`, () => {
    assert.equal(maskCapacity(c.mask), c.capacity);
    assert.equal(maskDescription(c.mask), c.description);
  });
}

for (const c of maskDoc.format) {
  test(`forms/mask format — ${JSON.stringify(c.mask)} <- ${JSON.stringify(c.raw)}`, () => {
    assert.equal(maskFormat(c.mask, c.raw), c.display);
    assert.equal(maskExtract(c.mask, c.display), c.extract);
  });
}

for (const c of maskDoc.extract) {
  test(`forms/mask extract — ${JSON.stringify(c.mask)} <- ${JSON.stringify(c.text)}`, () => {
    assert.equal(maskExtract(c.mask, c.text), c.raw);
  });
}

for (const c of maskDoc.edits) {
  test(`forms/mask edit — ${c.name}`, () => {
    const got = maskEdit(c.mask, c.prev, c.selStart, c.selEnd, c.insert);
    assert.deepEqual(
      { display: got.display, caret: got.caret, raw: got.raw, complete: got.complete },
      c.expect,
    );
  });
}

test("forms/mask — bind receives the UNMASKED value, always", () => {
  for (const c of maskDoc.edits) {
    assert.equal(maskExtract(c.mask, c.expect.display), c.expect.raw,
      `${c.name}: display and raw disagree`);
  }
});

// ── phone.json ────────────────────────────────────────────────────────────────────────

type PhoneDoc = {
  flags: { iso: string; flag: string }[];
  parse: {
    input: string; defaultCountry: string | null;
    expect: { e164: string; national: string; country: string | null;
              dialCode: string; nsn: string; valid: boolean };
  }[];
};
const phoneDoc = load<PhoneDoc>("phone.json");
assert.ok(phoneDoc.parse.length >= 30, "forms/phone parse corpus is suspiciously small");

for (const c of phoneDoc.flags) {
  test(`forms/phone flag — ${c.iso}`, () => { assert.equal(formsFlag(c.iso), c.flag); });
}

for (const c of phoneDoc.parse) {
  test(`forms/phone parse — ${JSON.stringify(c.input)} @ ${c.defaultCountry ?? "-"}`, () => {
    const got = phoneParse(c.input, c.defaultCountry);
    assert.deepEqual(
      { e164: got.e164, national: got.national, country: got.country,
        dialCode: got.dialCode, nsn: got.nsn, valid: got.valid },
      c.expect,
    );
  });
}

test("forms/phone — a valid parse round-trips through its own E.164", () => {
  for (const c of phoneDoc.parse) {
    if (!c.expect.valid) continue;
    const again = phoneParse(c.expect.e164, null);
    assert.equal(again.e164, c.expect.e164, `${c.input} did not round-trip`);
    assert.equal(again.valid, true, `${c.input} lost validity on round-trip`);
  }
});

// ── daterange.json ────────────────────────────────────────────────────────────────────

type DateDoc = {
  grid: { year: number; month: number; firstWeekday: number;
          expect: { days: number; leading: number; weeks: number } }[];
  dayNumbers: { date: string; day: number | null }[];
  dstAdjacency: { from: string; to: string; delta: number; next: string }[];
  selectable: { config: { min?: string; max?: string; disabledDates?: string[] };
                date: string; expect: { ok: boolean; reason: string } }[];
  folds: {
    name: string;
    config: { range?: boolean; min?: string; max?: string; disabledDates?: string[]; month?: string };
    events: DateFoldEvent[];
    expect: { start: string | null; end: string | null; month: string;
              complete: boolean; reason: string; fired: string[] }[];
  }[];
};
const dateDoc = load<DateDoc>("daterange.json");
assert.ok(dateDoc.folds.length >= 10, "forms/daterange fold corpus is suspiciously small");

for (const c of dateDoc.grid) {
  test(`forms/daterange grid — ${c.year}-${c.month} fw${c.firstWeekday}`, () => {
    assert.deepEqual({ ...formsMonthGrid(c.year, c.month, c.firstWeekday) }, c.expect);
  });
}

for (const c of dateDoc.dayNumbers) {
  test(`forms/daterange dayNumber — ${c.date}`, () => {
    assert.equal(formsDayNumber(c.date), c.day);
    if (c.day !== null) assert.equal(formsDateFromDay(c.day), c.date);
  });
}

for (const c of dateDoc.dstAdjacency) {
  test(`forms/daterange DST — ${c.from} -> ${c.to} is exactly one day`, () => {
    assert.equal((formsDayNumber(c.to) ?? 0) - (formsDayNumber(c.from) ?? 0), c.delta);
    assert.equal(formsDateFromDay((formsDayNumber(c.from) ?? 0) + 1), c.next);
  });
}

for (const c of dateDoc.selectable) {
  test(`forms/daterange selectable — ${c.date}`, () => {
    assert.deepEqual({ ...formsSelectable(c.config, c.date) }, c.expect);
  });
}

for (const c of dateDoc.folds) {
  test(`forms/daterange fold — ${c.name}`, () => {
    const steps = formsDateFold(c.config, c.events);
    assert.equal(steps.length, c.expect.length);
    for (let i = 0; i < steps.length; i += 1) {
      assert.deepEqual(
        { start: steps[i]!.start, end: steps[i]!.end, month: steps[i]!.month,
          complete: steps[i]!.complete, reason: steps[i]!.reason, fired: [...steps[i]!.fired] },
        c.expect[i], `step ${i}`,
      );
    }
  });
}

// ── validation.json ───────────────────────────────────────────────────────────────────

type ValidationDoc = {
  rules: { rule: string; arg: string; value: string; pattern: string; expect: boolean }[];
  fieldErrors: { field: FormFieldState; expect: string }[];
  aggregate: { name: string; fields: FormFieldState[];
               expect: { valid: boolean; dirty: boolean; errors: Record<string, string>; invalid: string[] } }[];
  submit: { name: string; state: { submitting: boolean; disabled: boolean };
            fields: FormFieldState[]; expect: Record<string, unknown> }[];
};
const validationDoc = load<ValidationDoc>("validation.json");
assert.ok(validationDoc.rules.length >= 15, "forms/validation rules corpus is suspiciously small");

for (const c of validationDoc.rules) {
  test(`forms/validation rule — ${c.rule}(${c.arg}) <- ${JSON.stringify(c.value)}`, () => {
    assert.equal(formsRule(c.rule, c.arg, c.value, c.pattern), c.expect);
  });
}

for (const [i, c] of validationDoc.fieldErrors.entries()) {
  test(`forms/validation fieldError — ${i} ${JSON.stringify(c.field.value ?? "")}`, () => {
    assert.equal(formsFieldError(c.field), c.expect);
  });
}

for (const c of validationDoc.aggregate) {
  test(`forms/validation aggregate — ${c.name}`, () => {
    const got = formsAggregate(c.fields);
    assert.deepEqual(
      { valid: got.valid, dirty: got.dirty, errors: got.errors, invalid: [...got.invalid] },
      c.expect,
    );
  });
}

for (const c of validationDoc.submit) {
  test(`forms/validation submit — ${c.name}`, () => {
    const got = formsSubmit(c.state, c.fields) as unknown as Record<string, unknown>;
    for (const [k, v] of Object.entries(c.expect)) {
      if (Array.isArray(v)) assert.deepEqual([...(got[k] as unknown[])], v, k);
      else assert.equal(got[k], v, k);
    }
  });
}

// ── composites.json ───────────────────────────────────────────────────────────────────

type CompositesDoc = {
  multiSelect: { selected: string[]; value: string; max: number;
                 expect: { selected: string[]; changed: boolean; reason: string } }[];
  multiSelectAnnouncement: { count: number; max: number; expect: string }[];
  tagsAdd: { tags: string[]; text: string;
             config: { separator?: string; max?: number; validate?: string };
             expect: { tags: string[]; added: string[]; rejected: { tag: string; reason: string }[] } }[];
  tagsBackspace: { tags: string[]; query: string;
                   expect: { tags: string[]; removed: string | null } }[];
  tagsAnnouncement: { kind: "add" | "remove"; tag: string; count: number; expect: string }[];
};
const composites = load<CompositesDoc>("composites.json");

for (const [i, c] of composites.multiSelect.entries()) {
  test(`forms/composites multiSelectToggle — ${i} ${c.value}`, () => {
    const got = multiSelectToggle(c.selected, c.value, c.max);
    assert.deepEqual({ selected: [...got.selected], changed: got.changed, reason: got.reason }, c.expect);
  });
}

for (const c of composites.multiSelectAnnouncement) {
  test(`forms/composites multiSelectAnnouncement — ${c.count}/${c.max}`, () => {
    assert.equal(multiSelectAnnouncement(c.count, c.max), c.expect);
  });
}

for (const [i, c] of composites.tagsAdd.entries()) {
  test(`forms/composites tagsAdd — ${i} ${JSON.stringify(c.text)}`, () => {
    const got = tagsAdd(c.tags, c.text, c.config);
    assert.deepEqual(
      { tags: [...got.tags], added: [...got.added], rejected: got.rejected.map((r) => ({ ...r })) },
      c.expect,
    );
  });
}

for (const [i, c] of composites.tagsBackspace.entries()) {
  test(`forms/composites tagsBackspace — ${i}`, () => {
    const got = tagsBackspace(c.tags, c.query);
    assert.deepEqual({ tags: [...got.tags], removed: got.removed }, c.expect);
  });
}

for (const c of composites.tagsAnnouncement) {
  test(`forms/composites tagsAnnouncement — ${c.kind} ${c.tag}`, () => {
    assert.equal(tagsAnnouncement(c.kind, c.tag, c.count), c.expect);
  });
}
